File: src/compiler/check.ts
import * as ts from "typescript";
import { convertDiagnostics } from "./diagnostics";
import { Diagnostic } from "./types";
import { host, compilerOptions } from "./typescript";

export type CheckRequest = { moduleIdentifiers: Array<string> };

export type CheckResponse = {
	diagnostics: { [key: string]: Array<Diagnostic> };
};

export let check = (request: CheckRequest): CheckResponse => {
	// Create a typescript program.
	let program = ts.createProgram({
		rootNames: [...request.moduleIdentifiers],
		options: compilerOptions,
		host,
	});

	// Get the diagnostics and convert them.
	let diagnostics = convertDiagnostics([
		...program.getConfigFileParsingDiagnostics(),
		...program.getOptionsDiagnostics(),
		...program.getGlobalDiagnostics(),
		...program.getDeclarationDiagnostics(),
		...program.getSyntacticDiagnostics(),
		...program.getSemanticDiagnostics(),
	]);

	return {
		diagnostics,
	};
};

File: src/compiler/check.rs
use super::{Diagnostic, ModuleIdentifier, Request, Response};
use crate::Cli;
use anyhow::{bail, Result};
use std::collections::BTreeMap;

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckRequest {
	pub module_identifiers: Vec<ModuleIdentifier>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckResponse {
	pub diagnostics: BTreeMap<ModuleIdentifier, Vec<Diagnostic>>,
}

impl Cli {
	/// Get all diagnostics for a package.
	pub async fn check(
		&self,
		module_identifiers: Vec<ModuleIdentifier>,
	) -> Result<BTreeMap<ModuleIdentifier, Vec<Diagnostic>>> {
		// Create the request.
		let request = Request::Check(CheckRequest { module_identifiers });

		// Send the request and receive the response.
		let response = self.request(request).await?;

		// Get the response.
		let response = match response {
			Response::Check(response) => response,
			_ => bail!("Unexpected response type."),
		};

		Ok(response.diagnostics)
	}
}
